compiler/rustc_mir_transform/src/check_pointers.rs

```diff
@@ -194,8 +193,24 @@ impl<'a, 'tcx> Visitor<'tcx> for PointerFinder<'a, 'tcx> {
             return;
         }
 
-        let pointee_ty =
-            pointer_ty.builtin_deref(true).expect("no builtin_deref for an raw pointer");
+        // If we see a borrow of a field projection, we want to pass the field Ty to the
+        // check and not the pointee Ty.
+        let pointee_ty = match self.field_projection_mode {
+            BorrowedFieldProjectionMode::FollowProjections
+                if matches!(
+                    context,
+                    PlaceContext::NonMutatingUse(NonMutatingUseContext::SharedBorrow)
+                        | PlaceContext::MutatingUse(MutatingUseContext::Borrow)
+                ) =>
+            {
+                if let Some(PlaceElem::Field(_, ty)) = place.projection.last() {
+                    *ty
+                } else {
+                    pointer_ty.builtin_deref(true).expect("no builtin_deref for an raw pointer")
```

Comment:
_Originally_ (when I first wrote this method) I thought we actually cared about whether or not the pointer involved in the Deref projection itself. That became not true pretty quickly and now with this PR it's _really_ not true.

So I feel like all of this logic should be based on `place.ty()` now, and this checking of the last field projection is just a partial implementation of that method's logic. Do you agree?

Reply:
I have to say I am not 100% sure I follow, but I agree that we should be always based on `place.ty()` and abstract from there. I reworked the code to make it a bit clearer, and the special case of deref projections for borrows is now just a special case. I put it into a separate commit for now, so that it becomes clearer.

Reply:
The change that you made here isn't what I was thinking of. The sketchy code is not in the assumption above that the type of the Pointer is the type of the Local; that assumption is easily justified by local reasoning. We're creating a Place without projections which is just a Local.

The sketchy part is that you are special-casing computing the pointee type based on checking if the last projection is a field projection. If another projection is added that changes the type, this logic becomes incomplete in a subtle way. I believe the purpose of this code is to compute the type of the place in question, and that's what Place::ty is for. 

Am I mistaken?

Reply:
Oke I think I got your point now, thanks for the explanation! I updated the code and hope I understood it correctly :)